python/thirdweb-ai/src/thirdweb_ai/services/storage.py

```diff
@@ -0,0 +1,176 @@
+import asyncio
+import hashlib
+import json
+import mimetypes
+import os
+from collections.abc import AsyncGenerator
+from dataclasses import asdict, is_dataclass
+from io import BytesIO
+from pathlib import Path
+from typing import Annotated, Any
+
+import httpx
+from pydantic import BaseModel
+
+from thirdweb_ai.services.service import Service
+from thirdweb_ai.tools.tool import tool
+
+
+async def async_read_file_chunks(file_path: str | Path, chunk_size: int = 8192) -> AsyncGenerator[bytes, None]:
+    """Read file in chunks asynchronously to avoid loading entire file into memory."""
+    async with asyncio.Lock():
+        path_obj = Path(file_path) if isinstance(file_path, str) else file_path
+        with path_obj.open("rb") as f:
+            while chunk := f.read(chunk_size):
+                yield chunk
+
+
+class Storage(Service):
+    def __init__(self, secret_key: str):
+        super().__init__(base_url="https://storage.thirdweb.com/ipfs", secret_key=secret_key)
+        self.gateway_url = self._get_gateway_url()
+
+    def _get_gateway_url(self) -> str:
+        return hashlib.sha256(self.secret_key.encode()).hexdigest()[:32]
+
+    @tool(description="Fetch content from IPFS by hash. Retrieves data stored on IPFS using the thirdweb gateway.")
+    def fetch_ipfs_content(
+        self,
+        ipfs_hash: Annotated[
+            str, "The IPFS hash/URI to fetch content from (e.g., 'ipfs://QmXyZ...'). Must start with 'ipfs://'."
+        ],
+    ) -> dict[str, Any]:
+        if not ipfs_hash.startswith("ipfs://"):
+            return {"error": "Invalid IPFS hash"}
+
+        ipfs_hash = ipfs_hash.removeprefix("ipfs://")
+        path = f"https://{self.gateway_url}.ipfscdn.io/ipfs/{ipfs_hash}"
```

Comment:
nit: would move the gateway hostname to the class constructor, so we can swap it with different environment for testing later